lib/kaitai/struct/struct.rb

```diff
@@ -531,6 +531,22 @@ def to_signed(x, mask)
   def self.format_hex(bytes)
     bytes.unpack('H*')[0].gsub(/(..)/, '\1 ').chop
   end
+
+  ###
+  # Guess about given args being byte arrays most likely.
+  # <p>
+  # There's no way to know for sure, but {@code Encoding::ASCII_8BIT} is a special encoding that is
+  # usually used for a byte array(/string), not a character string. For those reasons, that encoding
+  # is NOT planned to be allowed for human readable texts by KS in general as well.
+  # </p>
+  # @param args [...] Something to check.
+  # @see <a href="https://ruby-doc.org/core-3.0.0/Encoding.html">Encoding</a>
+  # @see <a href="https://github.com/kaitai-io/kaitai_struct/issues/116">List of supported encodings</a>
+  #
+  def self.is_byte_array?(*args)
+    found = args.select { |arg| arg.is_a?(String) and (arg.encoding == Encoding::ASCII_8BIT) }
+    found.length == args.length
+  end
```

Comment:
This function is unnecessarily too specific. I'd simplify this to one argument only:

```suggestion
  def self.is_byte_array?(arg)
    arg.is_a?(String) and (arg.encoding == Encoding::ASCII_8BIT)
  end
```

Then instead of

https://github.com/kaitai-io/kaitai_struct_ruby_runtime/blob/ae26b02793034b6018a1a31135ed82674abd6224/lib/kaitai/struct/struct.rb#L588

let's do

```ruby
    if Stream.is_byte_array?(expected) and Stream.is_byte_array?(actual)
```

or eventually

```ruby
    if [expected, actual].all? { |v| Stream.is_byte_array?(v) }
```

if you prefer. I think this is more idiomatic and also makes the method `Stream.is_byte_array?` more generic and reusable.

Reply:
I still believe it makes more sense to keep the caller short and simple. So, any interest do merge my implementation?

Reply:
> I still believe it makes more sense to keep the caller short and simple.

I don't think your implementation does this. Are you sure that this

```ruby
  def self.is_byte_array?(*args)
    found = args.select { |arg| arg.is_a?(String) and (arg.encoding == Encoding::ASCII_8BIT) }
    found.length == args.length
  end
```
```ruby
    if Stream.is_byte_array?(expected, actual)
```

is simpler than this?

```ruby
  def self.is_byte_array?(arg)
    arg.is_a?(String) and (arg.encoding == Encoding::ASCII_8BIT)
  end
```
```ruby
    if [expected, actual].all? { |v| Stream.is_byte_array?(v) }
```

Functions should be short, focus on just one thing and do it well. This ensures that they are easily reusable. When you need to assemble complex tasks, you just build it from these basic building blocks, not write a long function doing too many things (from scratch or using copy-paste, which is obviously bad).

I don't see why you should write another function only if you need a different "operation" than `.all?` (which is reproduced in your implementation as `found.length == args.length`; and yeah, it is actually the logical `AND` operation). What if I want to find out whether _at least one_ item out of `[expected, actual]` is a byte array (i.e. the `OR` operation)? Seems legitimate to me, I don't see how the `AND` operation has a privileged position over others.

The first time I saw this

```ruby
    if Stream.is_byte_array?(expected, actual)
```

I had no idea what the `Stream.is_byte_array?` actually does for multiple arguments - is it [`Array.all?`](https://docs.ruby-lang.org/en/3.0.0/Array.html#method-i-all-3F) or [`Array.any?`](https://docs.ruby-lang.org/en/3.0.0/Array.html#method-i-any-3F)? `is_byte_array` contains a singular verb **`is`** (infinitive _to be_), so this also doesn't say anything how the function will behave for more args.

---

> So, any interest do merge my implementation?

There sure is, but I generally won't approve anything that I'm convinced can be done better. In theory, I could merge PRs as they are submitted (even if they are dysfunctional) and _then_ do a few commits afterwards to reach the state that I consider best (and some contributors would certainly prefer this), but I consider that

1. disrespectful towards the PR authors, because I'd massively rework something that they put their effort in without them even knowing;
2. not beneficial for the PR authors at all; the PR review process is a unique chance for contributors to learn something, ask questions, get to know the best practices, etc. - I'd take all this away;
3. inefficient, because the PR author have surely done their research so they know why they did certain things, and I'd need to redo their effort instead of simply asking them;
4. a dangerous precedent that contributors can always just throw half-done dog excrements at me and I will even always thank them for it, while knowing that there will be at least several hours of work for me and I would hardly have time on anything else.

Reply:
> There sure is, but I generally won't approve anything that I'm
> convinced can be done better.[...]

It's obvious that you simply don't like my implementation, while your arguments don't hold too much:

Your call-site is pretty complex, especially for people new to ruby, while mine provides an easy to understand name of what is done. Your call-site has simply more chars than mine. Your call-site is different for the only currently known use-cases of checking one and multiple given objects, while mine is always the same. Because your call site is different for those cases, it would result in copy&paste at different places. Your argument with AND vs. possible OR is premature optimization, while using AND is needed right now.

So instead of forcing me with pretty weak arguments to follow your implementation in the end anyway, you should have simply changed and merged the PR already on your own. The only thing you are currently achieving is wasting my time, yours and making me angry, because of  that discussion. Really, you are trying to tell me what functions are to be used for and at the same time reverted a PR WITHOUT re-opening the corresponding issue as well to document that the problem is NOT fixed anymore?! So much for the basics...

It's perfectly fine that you don't like my approach, but yours is not SUPERIOR, only DIFFERENT.

Reply:
> So instead of forcing me with pretty weak arguments to follow your implementation in the end anyway, you should have simply changed and merged the PR already on your own.

I don't understand the grumpy tone, but will do that.

> Really, you are trying to tell me what functions are to be used for and at the same time reverted a PR WITHOUT re-opening the corresponding issue as well to document that the problem is NOT fixed anymore?! So much for the basics...

Yeah, that's definitely an oversight. If you had told me before, I would've done it immediately.